oasislmf/pytools/pla/structure.py

```diff
@@ -0,0 +1,158 @@
+        count (int): number of remaning amplification IDs associated with
+            current event ID
+        cursor (int): position in buffer
+        valid_length (int): length of buffer in 8-bit chunks
+        event_count (numpy.ndarray): array of event ID-count pairs
+        amp_factor (numpy.ndarray): array of amplification ID-loss pairs
+        plafactors (numba.typed.typeddict.Dict): PLA factors dictionary
+
+    Returns:
+        event_id (int): current event ID
+        count (int): number of remaining amplification IDs associated with
+            current event ID
+        plafactors (numba.typed.typeddict.Dict): PLA factors dictionary
+    """
+    while cursor < valid_length:
+
+        if count == 0:
+            event_id = event_count[cursor]['event_id']
+            count = event_count[cursor]['count']
+            cursor += 1
+
+        else:
+            amplification_id = amp_factor[cursor]['amplification_id']
+            loss_factor = amp_factor[cursor]['factor']
+            plafactors[(event_id, amplification_id)] = loss_factor
+            cursor += 1
+            count -= 1
+
+    return event_id, count, plafactors
+
+
+def get_post_loss_amplification_factors(path):
+    """
+    Get Post Loss Amplification (PLA) factors mapped to event ID-item ID pair.
+
+    lossfactors.bin is binary file with layout:
+        reserved header (4-byte int),
+        event ID 1 (4-byte int), number of amplification IDs for event ID 1 (4-byte int),
+        amplification ID 1 (4-byte int), loss factor for amplification ID 1 (4-byte float),
+        ...
+        amplification ID n (4-byte int), loss factor for amplification ID n (4-byte float),
+        event ID 2 (4-byte int), number of amplification IDs for event ID 2 (4-byte int),
+        ...
+        event ID N (4-byte int), number of amplification IDs for event ID N (4-byte int),
+        amplification ID 1 (4-byte int), loss factor for amplification ID 1 (4-byte float),
+        ...
+        amplification ID n (4-byte int), loss factor for amplification ID n (4-byte float)
+
+    Args:
+        path (str): path to lossfactors.bin file
+
+    Returns:
+        plafactors (dict): event ID-item ID pairs mapped to amplification IDs
+    """
+    plafactors = Dict.empty(
+        key_type=types.UniTuple(types.int64, 2), value_type=types.float64
+    )
+
+    with open(os.path.join(path, LOSS_FACTORS_FILE_NAME), 'rb') as f:
+        factors_buffer = memoryview(bytearray(BUFFER_SIZE))
+        event_count = np.ndarray(
+            N_PAIRS, buffer=factors_buffer, dtype=event_count_dtype
+        )
+        amp_factor = np.ndarray(
+            N_PAIRS, buffer=factors_buffer, dtype=amp_factor_dtype
+        )
+        f.readinto1(factors_buffer[:FILE_HEADER_SIZE])   # Ignore first 4 bytes
+
+        cursor = 0
+        valid_buffer = 0
+        count = 0
+        event_id = 0
+        while True:
+            len_read = f.readinto1(factors_buffer[valid_buffer:])
```

Comment:
I think you need readinto here because readinto1 can return less that ask even if the file has more data.
then it would error if it cut in the middle of a line.
readinto read the number of byte asked unless the stream ends
readinto1 read from 1 up to the number of byte asked unless the stream ends (then it can return 0)